### PR TITLE
feat: optionally output a debug artifact on compile

### LINKED ISSUE #2279: Output debug files for external trazability of errors
### Problem

Right now only nargo can figure out during execution which assertion errored. This is due to the fact that we are collecting debug info during compilation and using it in nargo execute, but never outputting it for external usage.

### Happy Case

A new compile flag that outputs debug info along with a file map that allows resolving the locations in the debug info.

### Alternatives Considered

Instead of attaching the file map, pre resolving the locations in debug info. This has been superceded by embedding the source of the file since it can improve debugging tools.

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by sirasistant**
# Description

## Problem\*

Resolves #2279

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_driver/src/contract.rs [hunk 1]
```diff
@@ -47,6 +48,8 @@ pub struct ContractFunction {
 
     #[serde(serialize_with = "serialize_circuit", deserialize_with = "deserialize_circuit")]
     pub bytecode: Circuit,
+
+    pub debug: DebugInfo,
```
Comment (sirasistant) [inline crates/noirc_driver/src/contract.rs:52]: debug was present in `CompiledProgram` but not in `ContractFunction`

File: crates/nargo/src/ops/preprocess.rs [hunk 2]
```diff
@@ -42,7 +42,7 @@ pub fn preprocess_contract_function<B: ProofSystemCompiler>(
     include_keys: bool,
     common_reference_string: &[u8],
     func: ContractFunction,
-) -> Result<PreprocessedContractFunction, B::Error> {
+) -> Result<(PreprocessedContractFunction, DebugInfo), B::Error> {
```
Comment (sirasistant) [inline crates/nargo/src/ops/preprocess.rs:45]: `preprocess_program` did return a tuple of the preprocessed program and the corresponding debug_info, but it was missing for `preprocess_contract_function`

File: crates/nargo_cli/src/cli/compile_cmd.rs [hunk 3]
```diff
@@ -38,6 +41,10 @@ pub(crate) struct CompileCommand {
     #[arg(long)]
     include_keys: bool,
 
+    /// Output debug files
+    #[arg(long)]
+    output_debug: bool,
```
Comment (TomAFrench) [inline crates/nargo_cli/src/cli/compile_cmd.rs:46]:
```suggestion
    /// Output debug files
    #[arg(long, hide = true)]
    output_debug: bool,
```

There's a good chance that we're going to be making a bunch of breaking changes to debug info in the near future and this is only really needed by A3 atm so we shouldn't make this a fully supported feature.

File: crates/nargo/src/artifacts/debug.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+
+use fm::FileId;
+use noirc_errors::debug_info::DebugInfo;
+use noirc_frontend::hir::Context;
+
+#[derive(Debug, Serialize, Deserialize)]
```
Comment (kevaundray) [inline crates/nargo/src/artifacts/debug.rs:8]: Can you add a doc comment explaining this?

File: crates/nargo/src/artifacts/debug.rs [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+pub struct DebugFile {
+    pub source: String,
+    pub path: String,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
```
Comment (kevaundray) [inline crates/nargo/src/artifacts/debug.rs:14]: Can you add a doc comment explaining this?

File: crates/nargo/src/artifacts/debug.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use std::collections::{BTreeMap, BTreeSet};
+
+use fm::FileId;
+use noirc_errors::debug_info::DebugInfo;
+use noirc_frontend::hir::Context;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DebugFile {
+    pub source: String,
+    pub path: String,
```
Comment (kevaundray) [inline crates/nargo/src/artifacts/debug.rs:11]: Thinking out loud: should this be a String or perhaps a Path?

Comment (sirasistant) [reply crates/nargo/src/artifacts/debug.rs:11]: I think Path cannot be Deserialize or stored in a BtreeMap because it doesn't implement Sized

Comment (sirasistant) [reply crates/nargo/src/artifacts/debug.rs:11]: Oh, I think we can use PathBuf

Comment (sirasistant) [reply crates/nargo/src/artifacts/debug.rs:11]: yeah, we should use pathbuf

**Review by TomAFrench — CHANGES_REQUESTED** [review]
I'm currently making pretty large changes to these same files for #2164 which I think this would conflict pretty heavily with. I think it would be easier if those changes went in first and then I can merge them into here.

**Comment by kevaundray** [comment]
> I'm currently making pretty large changes to these same files for #2164 which I think this would conflict pretty heavily with. I think it would be easier if those changes went in first and then I can merge them into here.

@TomAFrench what is the corresponding PR which is blocking this one?

**Review by TomAFrench — APPROVED** [review]
I reviewed the PRs again and the overlap is smaller and less tangled than I initially thought so I don't have any objections for this going in first. Sorry for the delay @sirasistant!

**Pull request merged**